### PR TITLE
Bug 1816714: ovirt - add vnic profile id to platform

### PR DESCRIPTION
**Pull request by rgolangh**
The vnic profile that all the VM shares can now be configurable through
the platform.
The wizard will also ask to the user the choose one, in case there are multiple profiles per network.

```yaml
platform:
  ovirt:
    vnicProfileID: xxxxxxxx-xxxx-xxxx-xxxx-xxxxxxxxxxxx
```

### CODE REVIEW

**Comment by rgolangh** [comment]
@patrickdillon @wking PTAL

**Comment by Jmainguy** [comment]
code looks solid to my noob eyeballs, nice work

**Comment by abonillabeeche** [comment]
Is this fix already in a nightly build?

**Comment by sdodson** [comment]
> Is this fix already in a nightly build?

no, it wouldn't be in a build if it hasn't merged

**Comment by sdodson** [comment]
@Gal-Zaidman @sandrobonazzola can you please review the ovirt specific bits here while installer team takes on review of the generic components

**Comment by patrickdillon** [comment]
The CI failure indicates a problem when `ovirt_vnic_profile_id` is not set in the install-config: https://prow.svc.ci.openshift.org/view/gcs/origin-ci-test/pr-logs/pull/openshift_installer/3406/pull-ci-openshift-installer-master-e2e-ovirt/1470

I don't think `ovirt_vnic_profile_id` can be set to required in order to maintain backward compatibility. Previously a default value was being set here: https://github.com/openshift/installer/pull/3406/files#diff-4f2a1c1c86a37e0831ca92d07d3f6949L25

Can we assume that default if the variable is not set in the install config? Ideally that would be done outside of terraform and passed in tfvars. If you need to do it in terraform, you will need to set a default value for the variable--but (again) doing this in go and not terraform is easier/cleaner.

**Comment by rgolangh** [comment]
> The CI failure indicates a problem when `ovirt_vnic_profile_id` is not set in the install-config: https://prow.svc.ci.openshift.org/view/gcs/origin-ci-test/pr-logs/pull/openshift_installer/3406/pull-ci-openshift-installer-master-e2e-ovirt/1470
> 
> I don't think `ovirt_vnic_profile_id` can be set to required in order to maintain backward compatibility. Previously a default value was being set here: https://github.com/openshift/installer/pull/3406/files#diff-4f2a1c1c86a37e0831ca92d07d3f6949L25
> 
> Can we assume that default if the variable is not set in the install config? Ideally that would be done outside of terraform and passed in tfvars. If you need to do it in terraform, you will need to set a default value for the variable--but (again) doing this in go and not terraform is easier/cleaner.

The assumption is what cause the bug - there is no way to 'guess' the vnic profile  name and choosing the first one in the list is a poor choice. That heuristic is my terrible mistake.

Like you mentioned, getting the defaults from go is much more desirable and I'll look for a clean way to do it.

File: pkg/types/ovirt/defaults/platform.go [hunk 1]
```diff
@@ -1,11 +1,80 @@
 package defaults
 
 import (
+	"fmt"
+
+	"github.com/ovirt/go-ovirt"
+
+	ovirtinstallconfig "github.com/openshift/installer/pkg/asset/installconfig/ovirt"
```
Comment (patrickdillon) [inline pkg/types/ovirt/defaults/platform.go:8]: In general, we can't have `pkg/types` depend on `pkg/asset`. Defaults set in this pkg will need to be static.

File: pkg/types/ovirt/defaults/platform.go [hunk 2]
```diff
@@ -1,11 +1,80 @@
...
+
+func setDefaultNetwork(p *ovirt.Platform) {
+	if p.NetworkName == "" {
+		p.NetworkName = DefaultNetworkName
+	}
+}
+
+func setDefaultVNICProfile(p *ovirt.Platform, con *ovirtsdk.Connection) error {
+	if p.VNICProfileID == "" {
+		clusterResponse, err := con.SystemService().ClustersService().ClusterService(p.ClusterID).Get().Follow("networks").Send()
```
Comment (patrickdillon) [inline pkg/types/ovirt/defaults/platform.go:41]: As I said in my other comment, any defaults added here need to be static values. For one, I think commands like `create manifests` or `create ignition-configs` should be executable without access to the rhv api. API access like this in general should live in `pkg/asset` but I'm not aware of a similar example where default values are set.

Comment (patrickdillon) [reply pkg/types/ovirt/defaults/platform.go:41]: > ~~For one, I think commands like `create manifests` or `create ignition-configs` should be executable without access to the rhv api.~~

I was wrong about this, so I updated. There are examples in AWS where installconfig is validated with API connections.

File: pkg/types/defaults/installconfig.go [hunk 3]
```diff
@@ -81,6 +84,10 @@ func SetInstallConfigDefaults(c *types.InstallConfig) {
 		vspheredefaults.SetPlatformDefaults(c.Platform.VSphere, c)
 	case c.Platform.BareMetal != nil:
 		baremetaldefaults.SetPlatformDefaults(c.Platform.BareMetal, c)
+	case c.Platform.Ovirt != nil:
+		if err := ovirtdefaults.SetPlatformDefaults(c.Platform.Ovirt); err != nil {
+			logrus.Fatal("failed to set platform defaults for oVirt", err)
+		}
```
Comment (abhinavdahiya) [inline pkg/types/defaults/installconfig.go:90]: all defaults that are not static must be handled in assets and not here.

File: pkg/types/ovirt/defaults/platform.go (hunk 1 above)
Comment (abhinavdahiya) [reply pkg/types/ovirt/defaults/platform.go:8]: yes, not cloud imports in pkg/types

File: pkg/types/ovirt/platform.go [hunk 4]
```diff
@@ -10,6 +10,8 @@ type Platform struct {
 	// The target network of all the network interfaces of the nodes. Omitting defaults to ovirtmgmt
 	// network which is a default network for evert ovirt cluster.
 	NetworkName string `json:"ovirt_network_name,omitempty"`
+	//VNICProfileID defines the VNIC profile ID to use the the VM network interfaces.
+	VNICProfileID string `json:"ovirt_vnic_profile_id,omitempty"`
```
Comment (abhinavdahiya) [inline pkg/types/ovirt/platform.go:14]: please include `// +optional` as this field will be optional.

also details about the defaults when not specified.

File: pkg/asset/cluster/tfvars.go [hunk 5]
```diff
@@ -433,6 +433,23 @@ func (t *TerraformVariables) Generate(parents asset.Parents) error {
...
+			installConfig.Config.Platform.Ovirt.ClusterID,
+			installConfig.Config.Platform.Ovirt.NetworkName,
+			con)
+		if err != nil {
+			return errors.Wrapf(err, "failed to compute values for oVirt platform")
+		}
+		if len(profiles) != 1 {
+			return errors.Wrapf(err, "failed to compute values for oVirt platform, there are multiple profiles.")
+		}
+		installConfig.Config.Platform.Ovirt.VNICProfileID = profiles[0].MustId()
```
Comment (patrickdillon) [inline pkg/asset/cluster/tfvars.go:452]: Shouldn't this logic only be run if `installConfig.Config.Platform.Ovirt.VNICProfileID` is not set?

File: pkg/asset/cluster/tfvars.go [hunk 6]
```diff
@@ -433,6 +433,23 @@ func (t *TerraformVariables) Generate(parents asset.Parents) error {
...
+
+		profiles, err := ovirtconfig.FetchVNICProfileByClusterNetwork(
+			installConfig.Config.Platform.Ovirt.ClusterID,
+			installConfig.Config.Platform.Ovirt.NetworkName,
+			con)
+		if err != nil {
+			return errors.Wrapf(err, "failed to compute values for oVirt platform")
+		}
+		if len(profiles) != 1 {
+			return errors.Wrapf(err, "failed to compute values for oVirt platform, there are multiple profiles.")
```
Comment (patrickdillon) [inline pkg/asset/cluster/tfvars.go:450]: If i understand correctly this is different default behavior than what you're replacing, but that behavior would be considered a bug so this is ok, in my opinion.

Comment (rgolangh) [reply pkg/asset/cluster/tfvars.go:450]: right.

File: pkg/asset/cluster/tfvars.go (hunk 5 above)
Comment (rgolangh) [reply pkg/asset/cluster/tfvars.go:452]: right, I missed it during the refactoring. adding back

File: pkg/types/ovirt/platform.go (hunk 4 above)
Comment (rgolangh) [reply pkg/types/ovirt/platform.go]: done

**Comment by patrickdillon** [comment]
I will try to summarize Slack conversation:

It doesn't look like this will set VNIC profile ID for compute nodes.

ovirt MachinePools are being added in #3399
the VNIC profile ID will also need to be added to the machine provider

Is there a benefit from merging this before these changes? What is the effect when the changes happen on control plane but not compute?

**Comment by rgolangh** [comment]
> I will try to summarize Slack conversation:
> 
> It doesn't look like this will set VNIC profile ID for compute nodes.
> 
Vnic and network name are currently generic definition that goes into the template creation. Both masters and workers gets them because they use the same template. 
As far as what this bug is about, this solves them urgent problem.

> ovirt MachinePools are being added in #3399
> the VNIC profile ID will also need to be added to the machine provider
>

The cluster api provider still doesn't support customizing the nics of a machine.
I need to know now if it is acceptable to have a vnic id in the platform level and then override them in the machine config. If not then I need to add vnic into the machine pool ASAP
 
> Is there a benefit from merging this before these changes? What is the effect when the changes happen on control plane but not compute?

File: pkg/asset/installconfig/ovirt/validators.go [hunk 7]
```diff
@@ -37,3 +40,32 @@ func authenticated(c *Config) survey.Validator {
...
+	if err != nil {
+		return err
+	}
+	defer con.Close()
+
+	return validateVNICProfile(*c.Ovirt, con)
+}
+
+func validateVNICProfile(p ovirt.Platform, con *ovirtsdk.Connection) error {
+	if p.VNICProfileID == "" {
```
Comment (patrickdillon) [inline pkg/asset/installconfig/ovirt/validators.go:56]: This validation is providing a default value, but I think you are already doing that elsewhere. Validation should be checking that if a user provides a value then it is valid.

**Comment by rgolangh** [comment]
@patrickdillon it really seems that adding this to machine spec makes more sense. So I added https://github.com/openshift/cluster-api-provider-ovirt/pull/45 and this will find its way into #3399 
After that I'll move `vnic_profile_id` from platform and into machine pool

File: pkg/asset/installconfig/ovirt/validators.go (hunk 7 above)
Comment (patrickdillon) [reply pkg/asset/installconfig/ovirt/validators.go:56]: The way I am reading this code:
if user provided a vnic profile ID: do nothing (it is valid)
if user did not provide a vnic profile ID: see if you can determine a default or else return an error

What I was expecting:
if user provided a vnic profile ID: validate (use API to determine if user provided value is correct)
if user did not provide: do nothing (default will be assigned later)

Does that make sense? I could definitely have mistaken assumptions here.

Comment (rgolangh) [reply pkg/asset/installconfig/ovirt/validators.go:56]: ack

**Pull request merged**